core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java

```diff
@@ -128,27 +139,150 @@ public void start() {
          Function<? super Publisher<CacheEntry<K, V>>, ? extends CompletionStage<R>> collator,
          Function<? super Publisher<R>, ? extends CompletionStage<R>> finalizer) {
       if (keysToInclude != null) {
-         return handleSpecificEntries(parallelPublisher, keysToInclude, keysToExclude, deliveryGuarantee, collator, finalizer);
+         return handleSpecificEntries(parallelPublisher, keysToInclude, keysToExclude, includeLoader, deliveryGuarantee,
+               collator, finalizer);
       }
 
-      AdvancedCache<K, V> cache = getCacheWithFlags(includeLoader);
+      CacheSet<CacheEntry<K, V>> entrySet = getEntrySet(includeLoader);
 
       // We have to cast to Function, since we can't cast our inner generic
       Function<CacheEntry<K, V>, K> toKeyFunction = (Function) StreamMarshalling.entryToKeyFunction();
       switch (deliveryGuarantee) {
          case AT_MOST_ONCE:
-            CompletionStage<R> stage = atMostOnce(parallelPublisher, cache.cacheEntrySet(), keysToExclude, toKeyFunction,
+            CompletionStage<R> stage = atMostOnce(parallelPublisher, entrySet, keysToExclude, toKeyFunction,
                   segments, collator, finalizer);
             return stage.thenApply(ignoreSegmentsFunction());
          case AT_LEAST_ONCE:
-            return atLeastOnce(parallelPublisher, cache.cacheEntrySet(), keysToExclude, toKeyFunction, segments, collator, finalizer);
+            return atLeastOnce(parallelPublisher, entrySet, keysToExclude, toKeyFunction, segments, collator, finalizer);
          case EXACTLY_ONCE:
-            return exactlyOnce(parallelPublisher, cache.cacheEntrySet(), keysToExclude, toKeyFunction, segments, collator, finalizer);
+            return exactlyOnce(parallelPublisher, entrySet, keysToExclude, toKeyFunction, segments, collator, finalizer);
          default:
             throw new UnsupportedOperationException("Unsupported delivery guarantee: " + deliveryGuarantee);
       }
    }
 
+   @Override
+   public <R> SegmentAwarePublisher<R> keyPublisher(IntSet segments, Set<K> keysToInclude,
+         Set<K> keysToExclude, boolean includeLoader, DeliveryGuarantee deliveryGuarantee,
+         Function<? super Publisher<K>, ? extends Publisher<R>> transformer) {
+      if (keysToInclude != null) {
+         AdvancedCache<K, V> cache = getCache(deliveryGuarantee, includeLoader);
+         return specificKeyPublisher(segments, keysToInclude, keyFlowable -> keyFlowable.filter(cache::containsKey),
+               transformer);
+      }
+      return new SegmentAwarePublisherImpl<>(segments, getKeySet(includeLoader), Function.identity(),
+            keysToExclude, deliveryGuarantee, transformer);
+   }
+
+   Flowable<CacheEntry<K, V>> filterEntries(AdvancedCache<K, V> cacheToUse, Flowable<K> entryFlowable) {
+      return entryFlowable.map(k -> {
+         // TODO: technically this can block if it ends up going remote due to a rehash
+         CacheEntry<K, V> entry = cacheToUse.getCacheEntry(k);
```

Comment:
This PR?

Reply:
Yes, it should be. Oddly enough I found out exactly how to fix this today too :)